language/move-core/types/src/u256.rs

```diff
@@ -0,0 +1,376 @@
+    pub fn checked_mul(self, rhs: Self) -> Option<Self> {
+        self.0.checked_mul(rhs.0).map(Self)
+    }
+
+    /// Checked integer division. Computes self / rhs, returning None if rhs == 0.
+    pub fn checked_div(self, rhs: Self) -> Option<Self> {
+        self.0.checked_div(rhs.0).map(Self)
+    }
+
+    /// Checked integer remainder. Computes self % rhs, returning None if rhs == 0.
+    pub fn checked_rem(self, rhs: Self) -> Option<Self> {
+        self.0.checked_rem(rhs.0).map(Self)
+    }
+
+    /// Checked integer remainder. Computes self % rhs, returning None if rhs == 0.
+    pub fn checked_shl(self, rhs: u32) -> Option<Self> {
+        if rhs >= U256_NUM_BITS as u32 {
+            return None;
+        }
+        Some(Self(self.0.shl(rhs)))
+    }
+
+    /// Checked shift right. Computes self >> rhs, returning None if rhs is larger than or equal to the number of bits in self.
+    pub fn checked_shr(self, rhs: u32) -> Option<Self> {
+        if rhs >= U256_NUM_BITS as u32 {
+            return None;
+        }
+        Some(Self(self.0.shr(rhs)))
+    }
+
+    /// Downcast to a an unsigned value of type T
+    /// T must be at most u128
+    pub fn down_cast_lossy<T: std::convert::TryFrom<u128>>(self) -> T {
+        // Size of this type
+        let type_size = size_of::<T>();
+        // Maximum value for this type
+        let max_val: u128 = if type_size < 16 {
+            (1u128 << (NUM_BITS_PER_BYTE * type_size)) - 1u128
+        } else {
+            u128::MAX
+        };
+        // This should never fail
+        match T::try_from(self.0.low_u128() & max_val) {
+            Ok(w) => w,
+            Err(_) => panic!("Fatal! Downcast failed"),
+        }
+    }
+}
+
+impl From<u8> for U256 {
+    fn from(n: u8) -> Self {
+        U256(PrimitiveU256::from(n))
+    }
+}
+
+impl From<u16> for U256 {
+    fn from(n: u16) -> Self {
+        U256(PrimitiveU256::from(n))
+    }
+}
+
+impl From<u32> for U256 {
+    fn from(n: u32) -> Self {
+        U256(PrimitiveU256::from(n))
+    }
+}
+
+impl From<u64> for U256 {
+    fn from(n: u64) -> Self {
+        U256(PrimitiveU256::from(n))
+    }
+}
+
+impl From<u128> for U256 {
+    fn from(n: u128) -> Self {
+        U256(PrimitiveU256::from(n))
+    }
+}
+
+impl From<BigInt> for U256 {
+    fn from(n: BigInt) -> Self {
+        U256(PrimitiveU256::from_little_endian(&n.to_bytes_le().1))
+    }
+}
+
+impl From<&U256> for BigInt {
+    fn from(n: &U256) -> Self {
+        BigInt::from_bytes_le(Sign::Plus, &n.to_le_bytes())
+    }
+}
+
+impl From<&U256> for EthnumU256 {
+    fn from(n: &U256) -> EthnumU256 {
+        // TODO (ade): use better solution for conversion
+        // Currently using str because EthnumU256 can be little or big endian
+        let num_str = format!("{:X}", n.0);
+        // TODO (ade): remove expect()
+        EthnumU256::from_str_radix(&num_str, 16).expect("Cannot convert to U256")
+    }
+}
```

Comment:
Whats the reason for implementing these conversions? This inadvertently places these types in the public api and we may not want to do that

Reply:
`BigInt`: Conversions are done between BigInt and primitives in Move Model
`EthnumU256`: Prover uses EthnumU256 and needs conversions.

I agree the downside to the current approach is leaking types. I can move the conversions closer to the usage areas instead.

Reply:
Why can't the prover use this type directly?

Reply:
Where "this type" is the `U256` wrapper that you define in this file.

Reply:
Did the prover already have support for u256? or did this pr add it?

If you want to do that in a followup that's fine, just make sure you include a comment indicating that these deps and From impls will be removed shortly

Reply:
It used some u256, which I didn't realize till I had implemented something else for other components 
The sensible thing is to harmonize 
I'll mark this as a TODO and have a follow up PR

Reply:
Created this to track https://github.com/move-language/move/issues/597